File: jvm/physical-plan/src/main/kotlin/PhysicalPlan.kt
// Copyright 2020 Andy Grove
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.ballistacompute.physical

import org.ballistacompute.datatypes.RecordBatch
import org.ballistacompute.datatypes.Schema


/**
 * A physical plan represents an executable piece of code that will produce data.
 */
interface PhysicalPlan {

    fun schema(): Schema

    /**
     * Execute a physical plan and produce a series of record batches.
     */
    fun execute(): Sequence<RecordBatch>

    /**
     * Returns the children (inputs) of this physical plan. This method is used to enable use of the
     * visitor pattern to walk a query tree.
     */
    fun children(): List<PhysicalPlan>

    fun pretty(): String {
        return format(this)
    }

}

/** Format a logical plan in human-readable form */
private fun format(plan: PhysicalPlan, indent: Int = 0): String {
    val b = StringBuilder()
    0.until(indent).forEach { b.append("\t") }
    b.append(plan.toString()).append("\n")
    plan.children().forEach { b.append(format(it, indent+1)) }
    return b.toString()
}

File: rust/ballista/src/physical_plan.rs
// Copyright 2020 Andy Grove
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Ballista Physical Plan (Experimental).
//!
//! The physical plan is a serializable data structure describing how the plan will be executed.
//!
//! It differs from the logical plan in that it deals with specific implementations of operators
//! (e.g. SortMergeJoin versus BroadcastHashJoin) whereas the logical plan just deals with an
//! abstract concept of a join.
//!
//! The physical plan also accounts for partitioning and ordering of data between operators.

use crate::arrow::datatypes::Schema;

#[derive(Debug, Clone)]
pub enum PhysicalPlan {
    /// Projection.
    Project(ProjectPlan),
    /// Filter a.k.a predicate.
    Filter(FilterPlan),
    /// Take the first `limit` elements of the child's single output partition.
    GlobalLimit(GlobalLimitPlan),
    /// Limit to be applied to each partition.
    LocalLimit(LocalLimitPlan),
    /// Sort on one or more sorting expressions.
    Sort(SortPlan),
    /// Hash aggregate
    HashAggregate(HashAggregatePlan),
    /// Performs a hash join of two child relations by first shuffling the data using the join keys.
    ShuffledHashJoin(ShuffledHashJoinPlan),
    /// Performs a shuffle that will result in the desired partitioning.
    ShuffleExchange(ShuffleExchangePlan),
    /// Scans a partitioned data source
    FileScan(FileScanPlan),
}

#[derive(Debug, Clone)]
pub enum JoinType {
    Inner,
}

#[derive(Debug, Clone)]
pub enum BuildSide {
    BuildLeft,
    BuildRight,
}

#[derive(Debug, Clone)]
pub enum SortDirection {
    Ascending,
    Descending,
}

#[derive(Debug, Clone)]
pub enum NullOrdering {
    NullsFirst,
    NullsLast,
}

#[derive(Debug, Clone)]
pub enum Partitioning {
    UnknownPartitioning(usize),
    HashPartitioning(usize, Vec<Expression>),
}

#[derive(Debug, Clone)]
pub struct ProjectPlan {
    child: Box<PhysicalPlan>,
    projection: Vec<Expression>,
}

#[derive(Debug, Clone)]
pub struct FilterPlan {
    child: Box<PhysicalPlan>,
    filter: Box<Expression>,
}

#[derive(Debug, Clone)]
pub struct GlobalLimitPlan {
    child: Box<PhysicalPlan>,
    limit: usize,
}

#[derive(Debug, Clone)]
pub struct LocalLimitPlan {
    child: Box<PhysicalPlan>,
    limit: usize,
}

#[derive(Debug, Clone)]
pub struct FileScanPlan {
    projection: Option<Vec<usize>>,
    partition_filters: Option<Vec<Expression>>,
    data_filters: Option<Vec<Expression>>,
    output_schema: Box<Schema>,
}

#[derive(Debug, Clone)]
pub struct ShuffleExchangePlan {
    child: Box<PhysicalPlan>,
    output_partitioning: Partitioning,
}

#[derive(Debug, Clone)]
pub struct ShuffledHashJoinPlan {
    left_keys: Vec<Expression>,
    right_keys: Vec<Expression>,
    build_side: BuildSide,
    join_type: JoinType,
    left: Box<PhysicalPlan>,
    right: Box<PhysicalPlan>,
}

#[derive(Debug, Clone)]
pub struct SortOrder {
    child: Box<Expression>,
    direction: SortDirection,
    null_ordering: NullOrdering,
}

#[derive(Debug, Clone)]
pub struct SortPlan {
    sort_order: Vec<SortOrder>,
    child: Box<PhysicalPlan>,
}

#[derive(Debug, Clone)]
pub struct HashAggregatePlan {
    group_expr: Vec<Expression>,
    aggr_expr: Vec<Expression>,
    child: Box<PhysicalPlan>,
}

/// Physical expression
#[derive(Debug, Clone)]
pub enum Expression {
    Column(usize),
    //TODO: add all the expressions here or possibly re-use the expressions from the logical plan
}
